front/lib/api/files/client_executable.ts

```diff
@@ -277,3 +291,317 @@ export async function getClientExecutableFileContent(
+  auth: Authenticator,
+  {
+    fileId,
+    conversationId,
+    revertedByAgentConfigurationId,
+  }: {
+    fileId: string;
+    conversationId: ModelId;
+    revertedByAgentConfigurationId: string;
+  }
+): Promise<
+  Result<
+    { fileResource: FileResource; revertedContent: string },
+    { tracked: boolean; message: string }
+  >
+> {
+  const fileResource = await FileResource.fetchById(auth, fileId);
+  if (!fileResource) {
+    return new Err({ message: `File not found: ${fileId}`, tracked: true });
+  }
+
+  if (fileResource.contentType !== clientExecutableContentType) {
+    return new Err({
+      message: `File '${fileId}' is not a content creation file (content type: ${fileResource.contentType})`,
+      tracked: true,
+    });
+  }
+
+  const workspaceId = auth.getNonNullableWorkspace().id;
+
+  const fileActions = await getFileActions(auth, fileId, conversationId);
+
+  if (fileActions.length === 0) {
+    return new Err({
+      message: "No MCP actions found for this file",
+      tracked: true,
+    });
+  }
+
+  // Validate that the most recent action is not already a revert
+  const mostRecentAction = fileActions[fileActions.length - 1];
+
+  if (
+    mostRecentAction?.toolConfiguration.originalName ===
+    CREATE_CONTENT_CREATION_FILE_TOOL_NAME
+  ) {
+    return new Err({
+      message: "Last action is a create, cannot revert to a create",
+      tracked: false,
+    });
+  }
+
+  if (
+    mostRecentAction?.toolConfiguration.originalName ===
+    REVERT_CONTENT_CREATION_FILE_TOOL_NAME
+  ) {
+    return new Err({
+      message: "Last action is a revert, cannot revert twice in a row",
+      tracked: false,
+    });
+  }
+
+  let lastRevertAction: AgentMCPActionModel | null = null;
+  let createAction: AgentMCPActionModel | null = null;
+  let startIndex: number = -1;
+
+  // Search backwards through actions to find:
+  // - The most recent revert action (if any)
+  // - The create action (there should be exactly one)
+  for (let i = fileActions.length - 1; i >= 0; i--) {
+    const toolName = fileActions[i].toolConfiguration.originalName;
+
+    if (
+      toolName === REVERT_CONTENT_CREATION_FILE_TOOL_NAME &&
+      lastRevertAction === null
+    ) {
+      // Set the start index as soon as we find a revert action
+      if (startIndex === -1) {
+        startIndex = i;
+      }
+      lastRevertAction = fileActions[i];
+    }
+
+    if (toolName === CREATE_CONTENT_CREATION_FILE_TOOL_NAME) {
+      // Only set the start index if we haven't already found a revert action
+      if (startIndex === -1) {
+        startIndex = i;
+      }
+      createAction = fileActions[i];
+      break;
+    }
+  }
+
+  if (createAction === null) {
+    return new Err({
+      message: `No create action found for file '${fileId}'`,
+      tracked: false,
+    });
+  }
+
+  let startingContent: string;
+
+  // If there was a previous revert, use its output as our baseline
```

Comment:
Revert still doesn't work if the previous revert attempt was failure (= you asked to revert twice in a row and failed). We should use as baseline only when it's successful maybe?